### PR TITLE
feat!: Share `Extension`s under `Arc`s

### PR DESCRIPTION
**Pull request by aborgna-q**
Extensions are defined once and shared throughout `ExtensionRegistr`ies, `Package`s, and soon `Hugr`s (#1613).
This _write-once then share around_ is a good usecase for `Arc`s, specially since the definitions are mostly read and rarely cloned.

This is a requisite for #1613, to avoid cloning all extensions for each new hugr.

BREAKING CHANGE: `ExtensionRegistry` and `Package` now wrap `Extension`s in `Arc`s.

### CODE REVIEW

File: .github/workflows/python-wheels.yml [hunk 1]
```diff
@@ -74,4 +74,4 @@ jobs:
         uses: pypa/gh-action-pypi-publish@release/v1
         with:
           verbose: true
-          skip-existing: true
\ No newline at end of file
+          skip-existing: true
```
Comment (aborgna-q) [inline .github/workflows/python-wheels.yml:77]: Drive-by from the pre-commit checks.

File: hugr-core/src/extension/op_def.rs [hunk 2]
```diff
@@ -611,7 +611,7 @@ pub(super) mod test {
         assert_eq!(def.misc.len(), 1);
 
         let reg =
-            ExtensionRegistry::try_new([PRELUDE.to_owned(), EXTENSION.to_owned(), e]).unwrap();
+            ExtensionRegistry::try_new([PRELUDE.clone(), EXTENSION.clone(), e.into()]).unwrap();
```
Comment (aborgna-q) [inline hugr-core/src/extension/op_def.rs:614]: Note that for all of these `to_owned` was still valid.
I just think that `clone`-ing the `Arc` is more explicit now, that we can use that.

**Review by mark-koch — APPROVED** [review]
